### PR TITLE
Remove warnings from Avalonia.Base

### PR DESCRIPTION
**Pull request by MarchingCube**
## What does the pull request do?

Attempts to remove warnings from `Avalonia.Base` to improve codebase health. Currently we have far too many warnings that hide other problems (like member shadowing, unused fields).

### CODE REVIEW

File: src/Avalonia.Base/PropertyStore/IPriorityValueEntry.cs [hunk 1]
```diff
@@ -10,8 +7,6 @@ namespace Avalonia.PropertyStore
     /// </summary>
     internal interface IPriorityValueEntry : IValue
     {
-        BindingPriority Priority { get; }
```
Comment (MarchingCube) [inline src/Avalonia.Base/PropertyStore/IPriorityValueEntry.cs:13]: One of the hidden warnings - `IValue` already has `Priority`.

File: src/Avalonia.Styling/IStyledElement.cs [hunk 2]
```diff
@@ -17,11 +17,6 @@ public interface IStyledElement :
         /// </summary>
         event EventHandler Initialized;
 
-        /// <summary>
-        /// Raised when resources on the element are changed.
-        /// </summary>
-        event EventHandler<ResourcesChangedEventArgs> ResourcesChanged;
```
Comment (MarchingCube) [inline src/Avalonia.Styling/IStyledElement.cs:23]: Another case of shadowing.

**Comment by MarchingCube** [comment]
Followup for this would be to:

1. Enable nullable annotations in entire `Avalonia.Base`
2. Enable warnings as errors for CI builds to prevent breakage.

**Comment by grokys** [comment]
> Followup for this would be to:
> 
> 1. Enable nullable annotations in entire `Avalonia.Base`
> 2. Enable warnings as errors for CI builds to prevent breakage.

I've tried this and unfortunately it's not going to work. `Avalonia.Base` uses too many generics and there's currently no way using nullable annotations to express "`T` is nullable if a reference type, non-nullable if a value type", which we require. This means that warnings would be sprinkled about `Avalonia.Base` even if we were to do everything correctly.

**Comment by maxkatz6** [comment]
@grokys
You can mark generic type as nullable with 
[MaybeNullAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.maybenullattribute?view=netcore-3.1). For netstandard2.0 it's required to add this attribute to assembly or use ready one - https://github.com/manuelroemer/Nullable

**Comment by grokys** [comment]
I'm aware of `MaybeNullAttribute` (and yes we need to add it to Avalonia), but it won't fix the warnings because you can't apply it to a generic parameter.

For example of this limitation in the BCL, try using `EqualityComparer<T>.Default.Equals` with a reference type.

**Comment by maxkatz6** [comment]
Ah, yes. You can't apply it to generic parameter, but you can apply it to generic arguments and return value. It covers some possible cases in Avalonia.

**Comment by maxkatz6** [comment]
For instance, StyledPropertyBase from Avalonia.Base can use MaybeNull
```csharp
[return: MaybeNull]
public TValue GetDefaultValue(Type type)
{
    Contract.Requires<ArgumentNullException>(type != null);

    return GetMetadata(type).DefaultValue;
}
```

Where TValue is generic, and it isn't possible to use simple "?" with it.
And intellisence will understand it in expected way
image

File: src/Avalonia.Base/Data/BindingValue.cs [hunk 3]
```diff
@@ -190,7 +190,7 @@ public BindingValue<T> WithValue(T value)
         /// Gets the value of the binding value if present, otherwise the default value.
         /// </summary>
         /// <returns>The value.</returns>
-        public T GetValueOrDefault() => HasValue ? _value : default;
+        public T GetValueOrDefault() => HasValue ? _value : default!;
```
Comment (grokys) [inline src/Avalonia.Base/Data/BindingValue.cs:193]: I think these warnings would be better fixed with the `[return: MaybeNull]` attribute.

File: src/Avalonia.Base/ValueStore.cs [hunk 4]
```diff
@@ -162,7 +162,7 @@ public void ClearLocalValue<T>(StyledPropertyBase<T> property)
                         _sink.ValueChanged(new AvaloniaPropertyChangedEventArgs<T>(
                             _owner,
                             property,
-                            old,
+                            new Optional<T>(old),
```
Comment (grokys) [inline src/Avalonia.Base/ValueStore.cs:165]: This fixes the warning without using `!`. The warning is actually caused by a bug in Roslyn: https://github.com/dotnet/roslyn/issues/43225

**Review by grokys — APPROVED** [review]

**Pull request merged**